File: src/infrastructure/azure/pull-request-publisher.ts
import type { AzurePullRequest, CreatePullRequestInput } from '../../azure'
import type { Config } from '../config/config.models'
import type { GitContext } from '../git/git-context.models'

export interface PullRequestPublisher {
  publish(
    config: Config,
    context: GitContext,
    targets: string[],
    input: Pick<CreatePullRequestInput, 'title' | 'description' | 'workItemRefs'>,
    reviewerForTarget?: (target: string) => string
  ): Promise<Array<{ target: string; pullRequest: AzurePullRequest }>>
}

File: apps/dart/lib/src/features/describe/pull_request_publisher.dart
import '../../app/app_effect.dart';

abstract interface class PullRequestPublisher {
  AppEffect<List<PublishedPullRequest>> publish(
    List<String> targets,
    PullRequestDraft draft, {
    String Function(String target)? reviewerForTarget,
  });
}

final class PullRequestDraft {
  const PullRequestDraft({
    required this.title,
    required this.description,
    this.workItemIds = const [],
  });

  final String title;
  final String description;
  final List<String> workItemIds;
}

final class PublishedPullRequest {
  const PublishedPullRequest({
    required this.target,
    required this.id,
    this.url,
  });

  final String target;
  final int id;
  final String? url;
}
